File: cursed_tts/phonemes.py
"""
Full CMUdict-compatible ARPAbet phoneme inventory.

This covers all phonemes needed for general English TTS.
Stress markers (0, 1, 2) are stripped - we use stress-free phones.
"""

from typing import NamedTuple
import numpy as np


class PhonemeInfo(NamedTuple):
    """Metadata for a phoneme."""
    symbol: str
    ipa: str
    phoneme_type: str  # 'vowel', 'consonant', 'semivowel'
    example: str
    description: str


# Full CMUdict ARPAbet inventory (39 phonemes, stress-stripped)
# Organized by phoneme type for synthesis parameters
PHONEME_INVENTORY = {
    # === VOWELS (monophthongs) ===
    'AA': PhonemeInfo('AA', 'ɑ', 'vowel', 'odd', 'open back unrounded'),
    'AE': PhonemeInfo('AE', 'æ', 'vowel', 'at', 'near-open front unrounded'),
    'AH': PhonemeInfo('AH', 'ʌ', 'vowel', 'hut', 'open-mid back unrounded'),
    'AO': PhonemeInfo('AO', 'ɔ', 'vowel', 'ought', 'open-mid back rounded'),
    'EH': PhonemeInfo('EH', 'ɛ', 'vowel', 'ed', 'open-mid front unrounded'),
    'ER': PhonemeInfo('ER', 'ɝ', 'vowel', 'hurt', 'r-colored mid central'),
    'IH': PhonemeInfo('IH', 'ɪ', 'vowel', 'it', 'near-close front unrounded'),
    'IY': PhonemeInfo('IY', 'i', 'vowel', 'eat', 'close front unrounded'),
    'UH': PhonemeInfo('UH', 'ʊ', 'vowel', 'hood', 'near-close back rounded'),
    'UW': PhonemeInfo('UW', 'u', 'vowel', 'two', 'close back rounded'),
    
    # === VOWELS (diphthongs) ===
    'AW': PhonemeInfo('AW', 'aʊ', 'vowel', 'cow', 'diphthong'),
    'AY': PhonemeInfo('AY', 'aɪ', 'vowel', 'hide', 'diphthong'),
    'EY': PhonemeInfo('EY', 'eɪ', 'vowel', 'ate', 'diphthong'),
    'OW': PhonemeInfo('OW', 'oʊ', 'vowel', 'oat', 'diphthong'),
    'OY': PhonemeInfo('OY', 'ɔɪ', 'vowel', 'toy', 'diphthong'),
    
    # === CONSONANTS: Stops ===
    'P': PhonemeInfo('P', 'p', 'consonant', 'pea', 'voiceless bilabial stop'),
    'B': PhonemeInfo('B', 'b', 'consonant', 'bee', 'voiced bilabial stop'),
    'T': PhonemeInfo('T', 't', 'consonant', 'tea', 'voiceless alveolar stop'),
    'D': PhonemeInfo('D', 'd', 'consonant', 'dee', 'voiced alveolar stop'),
    'K': PhonemeInfo('K', 'k', 'consonant', 'key', 'voiceless velar stop'),
    'G': PhonemeInfo('G', 'ɡ', 'consonant', 'gee', 'voiced velar stop'),
    
    # === CONSONANTS: Affricates ===
    'CH': PhonemeInfo('CH', 'tʃ', 'consonant', 'cheese', 'voiceless postalveolar affricate'),
    'JH': PhonemeInfo('JH', 'dʒ', 'consonant', 'gee', 'voiced postalveolar affricate'),
    
    # === CONSONANTS: Fricatives ===
    'F': PhonemeInfo('F', 'f', 'consonant', 'fee', 'voiceless labiodental fricative'),
    'V': PhonemeInfo('V', 'v', 'consonant', 'vee', 'voiced labiodental fricative'),
    'TH': PhonemeInfo('TH', 'θ', 'consonant', 'thin', 'voiceless dental fricative'),
    'DH': PhonemeInfo('DH', 'ð', 'consonant', 'then', 'voiced dental fricative'),
    'S': PhonemeInfo('S', 's', 'consonant', 'sea', 'voiceless alveolar fricative'),
    'Z': PhonemeInfo('Z', 'z', 'consonant', 'zee', 'voiced alveolar fricative'),
    'SH': PhonemeInfo('SH', 'ʃ', 'consonant', 'she', 'voiceless postalveolar fricative'),
    'ZH': PhonemeInfo('ZH', 'ʒ', 'consonant', 'measure', 'voiced postalveolar fricative'),
    'HH': PhonemeInfo('HH', 'h', 'consonant', 'he', 'voiceless glottal fricative'),
    
    # === CONSONANTS: Nasals ===
    'M': PhonemeInfo('M', 'm', 'consonant', 'me', 'bilabial nasal'),
    'N': PhonemeInfo('N', 'n', 'consonant', 'knee', 'alveolar nasal'),
    'NG': PhonemeInfo('NG', 'ŋ', 'consonant', 'sing', 'velar nasal'),
    
    # === CONSONANTS: Liquids ===
    'L': PhonemeInfo('L', 'l', 'consonant', 'lee', 'alveolar lateral'),
    'R': PhonemeInfo('R', 'ɹ', 'consonant', 'ray', 'alveolar approximant'),
    
    # === SEMIVOWELS / Glides ===
    'W': PhonemeInfo('W', 'w', 'semivowel', 'way', 'labio-velar approximant'),
    'Y': PhonemeInfo('Y', 'j', 'semivowel', 'yes', 'palatal approximant'),
}

# Ordered list of phonemes (for MBON indexing)
PHONEME_LIST = list(PHONEME_INVENTORY.keys())
NUM_PHONEMES = len(PHONEME_LIST)

# Index mappings
PHONEME_TO_IDX = {p: i for i, p in enumerate(PHONEME_LIST)}
IDX_TO_PHONEME = {i: p for i, p in enumerate(PHONEME_LIST)}

# Phoneme categories for synthesis
VOWELS = {p for p, info in PHONEME_INVENTORY.items() if info.phoneme_type == 'vowel'}
CONSONANTS = {p for p, info in PHONEME_INVENTORY.items() if info.phoneme_type == 'consonant'}
SEMIVOWELS = {p for p, info in PHONEME_INVENTORY.items() if info.phoneme_type == 'semivowel'}

# Sub-categories for synthesis
STOPS = {'P', 'B', 'T', 'D', 'K', 'G'}
AFFRICATES = {'CH', 'JH'}
FRICATIVES = {'F', 'V', 'TH', 'DH', 'S', 'Z', 'SH', 'ZH', 'HH'}
NASALS = {'M', 'N', 'NG'}
LIQUIDS = {'L', 'R'}
VOICED_CONSONANTS = {'B', 'D', 'G', 'JH', 'V', 'DH', 'Z', 'ZH', 'M', 'N', 'NG', 'L', 'R'}
DIPHTHONGS = {'AW', 'AY', 'EY', 'OW', 'OY'}


def strip_stress(phoneme: str) -> str:
    """Strip stress marker (0, 1, 2) from a phoneme."""
    return phoneme.rstrip('012')


def phoneme_to_index(phoneme: str) -> int:
    """Convert a phoneme symbol to its MBON compartment index."""
    p = strip_stress(phoneme)
    if p not in PHONEME_TO_IDX:
        raise ValueError(f"Unknown phoneme: {phoneme} (stripped: {p}). Known: {PHONEME_LIST}")
    return PHONEME_TO_IDX[p]


def index_to_phoneme(idx: int) -> str:
    """Convert an MBON compartment index to its phoneme symbol."""
    if idx not in IDX_TO_PHONEME:
        raise ValueError(f"Invalid index: {idx}. Valid range: 0-{NUM_PHONEMES-1}")
    return IDX_TO_PHONEME[idx]


def get_phoneme_info(phoneme: str) -> PhonemeInfo:
    """Get metadata for a phoneme."""
    p = strip_stress(phoneme)
    if p not in PHONEME_INVENTORY:
        raise ValueError(f"Unknown phoneme: {phoneme}")
    return PHONEME_INVENTORY[p]


def is_vowel(phoneme: str) -> bool:
    """Check if a phoneme is a vowel."""
    return strip_stress(phoneme) in VOWELS


def is_consonant(phoneme: str) -> bool:
    """Check if a phoneme is a consonant."""
    return strip_stress(phoneme) in CONSONANTS


def is_voiced(phoneme: str) -> bool:
    """Check if a consonant is voiced."""
    p = strip_stress(phoneme)
    return p in VOICED_CONSONANTS or p in VOWELS or p in SEMIVOWELS


def print_inventory():
    """Print the phoneme inventory for reference."""
    print(f"\nPhoneme Inventory ({NUM_PHONEMES} phonemes):")
    print("-" * 70)
    print(f"{'Symbol':<8} {'IPA':<6} {'Type':<12} {'Example':<10} {'Description'}")
    print("-" * 70)
    
    current_type = None
    for symbol, info in PHONEME_INVENTORY.items():
        if info.phoneme_type != current_type:
            current_type = info.phoneme_type
            print(f"\n  === {current_type.upper()}S ===")
        print(f"{symbol:<8} {info.ipa:<6} {info.phoneme_type:<12} {info.example:<10} {info.description}")
    print()


if __name__ == "__main__":
    print_inventory()
    print(f"\nTotal: {NUM_PHONEMES} phonemes")
    print(f"Vowels: {len(VOWELS)}")
    print(f"Consonants: {len(CONSONANTS)}")
    print(f"Semivowels: {len(SEMIVOWELS)}")

File: webui/src/engine/phonemes.ts
import { PHONEME_LIST, PHONEME_TO_IDX, type Phoneme } from "./types";

export type PhonemeInfo = {
  symbol: Phoneme;
  ipa: string;
  kind: "vowel" | "consonant" | "semivowel";
  example: string;
  description: string;
};

export const PHONEME_INVENTORY: Record<Phoneme, PhonemeInfo> = {
  AA: { symbol: "AA", ipa: "ɑ", kind: "vowel", example: "odd", description: "open back unrounded" },
  AE: { symbol: "AE", ipa: "æ", kind: "vowel", example: "at", description: "near-open front unrounded" },
  AH: { symbol: "AH", ipa: "ʌ", kind: "vowel", example: "hut", description: "open-mid back unrounded" },
  AO: { symbol: "AO", ipa: "ɔ", kind: "vowel", example: "ought", description: "open-mid back rounded" },
  EH: { symbol: "EH", ipa: "ɛ", kind: "vowel", example: "ed", description: "open-mid front unrounded" },
  ER: { symbol: "ER", ipa: "ɝ", kind: "vowel", example: "hurt", description: "r-colored mid central" },
  IH: { symbol: "IH", ipa: "ɪ", kind: "vowel", example: "it", description: "near-close front unrounded" },
  IY: { symbol: "IY", ipa: "i", kind: "vowel", example: "eat", description: "close front unrounded" },
  UH: { symbol: "UH", ipa: "ʊ", kind: "vowel", example: "hood", description: "near-close back rounded" },
  UW: { symbol: "UW", ipa: "u", kind: "vowel", example: "two", description: "close back rounded" },
  AW: { symbol: "AW", ipa: "aʊ", kind: "vowel", example: "cow", description: "diphthong" },
  AY: { symbol: "AY", ipa: "aɪ", kind: "vowel", example: "hide", description: "diphthong" },
  EY: { symbol: "EY", ipa: "eɪ", kind: "vowel", example: "ate", description: "diphthong" },
  OW: { symbol: "OW", ipa: "oʊ", kind: "vowel", example: "oat", description: "diphthong" },
  OY: { symbol: "OY", ipa: "ɔɪ", kind: "vowel", example: "toy", description: "diphthong" },
  P: { symbol: "P", ipa: "p", kind: "consonant", example: "pea", description: "voiceless bilabial stop" },
  B: { symbol: "B", ipa: "b", kind: "consonant", example: "bee", description: "voiced bilabial stop" },
  T: { symbol: "T", ipa: "t", kind: "consonant", example: "tea", description: "voiceless alveolar stop" },
  D: { symbol: "D", ipa: "d", kind: "consonant", example: "dee", description: "voiced alveolar stop" },
  K: { symbol: "K", ipa: "k", kind: "consonant", example: "key", description: "voiceless velar stop" },
  G: { symbol: "G", ipa: "ɡ", kind: "consonant", example: "gee", description: "voiced velar stop" },
  CH: { symbol: "CH", ipa: "tʃ", kind: "consonant", example: "cheese", description: "voiceless postalveolar affricate" },
  JH: { symbol: "JH", ipa: "dʒ", kind: "consonant", example: "gee", description: "voiced postalveolar affricate" },
  F: { symbol: "F", ipa: "f", kind: "consonant", example: "fee", description: "voiceless labiodental fricative" },
  V: { symbol: "V", ipa: "v", kind: "consonant", example: "vee", description: "voiced labiodental fricative" },
  TH: { symbol: "TH", ipa: "θ", kind: "consonant", example: "thin", description: "voiceless dental fricative" },
  DH: { symbol: "DH", ipa: "ð", kind: "consonant", example: "then", description: "voiced dental fricative" },
  S: { symbol: "S", ipa: "s", kind: "consonant", example: "sea", description: "voiceless alveolar fricative" },
  Z: { symbol: "Z", ipa: "z", kind: "consonant", example: "zee", description: "voiced alveolar fricative" },
  SH: { symbol: "SH", ipa: "ʃ", kind: "consonant", example: "she", description: "voiceless postalveolar fricative" },
  ZH: { symbol: "ZH", ipa: "ʒ", kind: "consonant", example: "measure", description: "voiced postalveolar fricative" },
  HH: { symbol: "HH", ipa: "h", kind: "consonant", example: "he", description: "voiceless glottal fricative" },
  M: { symbol: "M", ipa: "m", kind: "consonant", example: "me", description: "bilabial nasal" },
  N: { symbol: "N", ipa: "n", kind: "consonant", example: "knee", description: "alveolar nasal" },
  NG: { symbol: "NG", ipa: "ŋ", kind: "consonant", example: "sing", description: "velar nasal" },
  L: { symbol: "L", ipa: "l", kind: "consonant", example: "lee", description: "alveolar lateral" },
  R: { symbol: "R", ipa: "ɹ", kind: "consonant", example: "ray", description: "alveolar approximant" },
  W: { symbol: "W", ipa: "w", kind: "semivowel", example: "way", description: "labio-velar approximant" },
  Y: { symbol: "Y", ipa: "j", kind: "semivowel", example: "yes", description: "palatal approximant" },
};

export const VOWELS = new Set(
  PHONEME_LIST.filter((p) => PHONEME_INVENTORY[p].kind === "vowel"),
);
export const CONSONANTS = new Set(
  PHONEME_LIST.filter((p) => PHONEME_INVENTORY[p].kind === "consonant"),
);
export const SEMIVOWELS = new Set(
  PHONEME_LIST.filter((p) => PHONEME_INVENTORY[p].kind === "semivowel"),
);
export const STOPS = new Set(["P", "B", "T", "D", "K", "G"]);
export const AFFRICATES = new Set(["CH", "JH"]);
export const FRICATIVES = new Set(["F", "V", "TH", "DH", "S", "Z", "SH", "ZH", "HH"]);
export const NASALS = new Set(["M", "N", "NG"]);
export const LIQUIDS = new Set(["L", "R"]);
export const VOICED_CONSONANTS = new Set([
  "B", "D", "G", "JH", "V", "DH", "Z", "ZH", "M", "N", "NG", "L", "R",
]);
export const DIPHTHONGS = new Set(["AW", "AY", "EY", "OW", "OY"]);

export function stripStress(phoneme: string): string {
  return phoneme.replace(/[012]$/, "");
}

export function asPhoneme(raw: string): Phoneme {
  const p = stripStress(raw.toUpperCase());
  if (!(p in PHONEME_TO_IDX)) {
    throw new Error(`Unknown phoneme: ${raw}`);
  }
  return p as Phoneme;
}

export function isPhoneme(raw: string): raw is Phoneme {
  return raw in PHONEME_TO_IDX;
}
